File: bot.js
const { Client, Intents, MessageEmbed } = require('discord.js');
const fs = require('fs');

const client = new Client({
  intents: [
    Intents.FLAGS.GUILDS,
    Intents.FLAGS.GUILD_MEMBERS,
    Intents.FLAGS.GUILD_BANS,
    Intents.FLAGS.GUILD_MESSAGES,
    Intents.FLAGS.DIRECT_MESSAGES
  ],
});

const allowedUserId = 'USER IDS'; // User IDs for moderators who can ban
const commandPrefix = '...'; //prefix
const bannedUsersFile = 'bannedUsers.json'; //banned users file for storing and accessing user DB of banned users
const appealEmail = 'appeals@bytelabs.site'; //Appeal email inserted here
const blockedWords = ['BAD WORD','BADWORD'].filter(word => word.trim() !== ''); // Badwords which are logged and moderated
const logChannelId = 'log channel'; // Logs all "badwords" sent in channels the bot can see and sends it to the channel

const bansPerPage = 10; // Number of bans to display per page

let bannedUsers = new Map();

// Load banned users from file
if (fs.existsSync(bannedUsersFile)) {
  try {
    const bannedUsersData = fs.readFileSync(bannedUsersFile, 'utf-8');
    bannedUsers = new Map(JSON.parse(bannedUsersData));
  } catch (error) {
    console.error('Error parsing banned users data:', error);
  }
}

function getUserIdFromMessage(message) {
  const mentionRegex = /^<@!?(\d+)>$/;
  const userIdRegex = /^(\d+)$/;

  if (mentionRegex.test(message)) {
    return mentionRegex.exec(message)[1];
  } else if (userIdRegex.test(message)) {
    return userIdRegex.exec(message)[1];
  } else {
    return null;
  }
}

function saveBannedUsersToFile() {
  const bannedUsersData = JSON.stringify(Array.from(bannedUsers));
  fs.writeFileSync(bannedUsersFile, bannedUsersData, 'utf-8');
}

function refreshBannedUsersCache() {
  bannedUsers = new Map(JSON.parse(fs.readFileSync(bannedUsersFile, 'utf-8')));
}

// Function to check if a name contains any of the blocked words
function containsBlockedWords(name) {
  const lowerCaseName = name.toLowerCase();
  return blockedWords.some((word) => lowerCaseName.includes(word));
}

async function unbanUserFromAllServers(userId) {
  const guilds = client.guilds.cache.values();
  for (const guild of guilds) {
    try {
      const bans = await guild.bans.fetch();
      const bannedUser = bans.find((ban) => ban.user.id === userId);
      if (bannedUser) {
        await guild.bans.remove(userId);
        console.log(`Unbanned user with ID ${userId} from ${guild.name}`);
      }
    } catch (error) {
      console.error(`Error unbanning user with ID ${userId} from ${guild.name}:`, error);
    }
  }
}

async function kickUnder7DaysOldMembers() {
  try {
    const guilds = client.guilds.cache.values();
    for (const guild of guilds) {
      try {
        const members = await guild.members.fetch();
        for (const member of members.values()) {
          const accountAgeInDays = (Date.now() - member.user.createdTimestamp) / (1000 * 60 * 60 * 24);
          if (accountAgeInDays < 7) {
            try {
              // Send a direct message to the user (optional)
              const dmChannel = await member.user.createDM();
              if (dmChannel) {
                const kickMessage =
                  `Hello ${member.user.username},\n` +
                  `You have been kicked from this server because your Discord account is less than 7 days old. ` +
                  `If you believe this is a mistake, please email your appeal to ${appealEmail}.`;
                await dmChannel.send(kickMessage);
              }
            } catch (error) {
              console.error('Error sending DM to kicked user:', error);
            }

            // Kick the user from the server
            try {
              await member.kick('Account age is less than 7 days');
              console.log(`Kicked ${member.user.tag} from ${guild.name}`);
            } catch (error) {
              if (error.code === 50013) {
                // If the bot lacks the necessary permissions, it will not attempt to kick the user.
                console.error(`Missing permissions to kick ${member.user.tag} from ${guild.name}`);
              } else {
                console.error(`Error kicking ${member.user.tag} from ${guild.name}:`, error);
              }
            }
          }
        }
      } catch (error) {
        console.error(`Error fetching members from ${guild.name}:`, error);
      }
    }
  } catch (error) {
    console.error('Error fetching guilds:', error);
  }
}

// Function to kick users with inappropriate usernames
async function kickUsersWithInappropriateNames() {
  try {
    const guilds = client.guilds.cache.values();
    for (const guild of guilds) {
      try {
        const members = await guild.members.fetch();
        for (const member of members.values()) {
          if (containsBlockedWords(member.user.username)) {
            try {
              // Send a direct message to the user
              const dmChannel = await member.user.createDM();
              if (dmChannel) {
                const kickMessage =
                  `Hello ${member.user.username},\n` +
                  `You have been kicked from ${guild.name} because your username or nickname contains inappropriate words. ` +
                  `If you believe this is a mistake or have any queries, please email ${appealEmail}.`;
                await dmChannel.send(kickMessage);
              }
            } catch (error) {
              console.error('Error sending DM to kicked user:', error);
            }

            // Kick the user from the server
            try {
              await member.kick('Inappropriate username detected');
              console.log(`Kicked ${member.user.tag} from ${guild.name} for having an inappropriate username.`);
            } catch (error) {
              if (error.code === 50013) {
                console.error(`Missing permissions to kick ${member.user.tag} from ${guild.name}`);
              } else {
                console.error(`Error kicking ${member.user.tag} from ${guild.name}:`, error);
              }
            }
          }
          if (member.nickname && containsBlockedWords(member.nickname)) {
            try {
              // Send a direct message to the user
              const dmChannel = await member.user.createDM();
              if (dmChannel) {
                const kickMessage =
                  `Hello ${member.user.username},\n` +
                  `You have been kicked from ${guild.name} because your username or nickname contains inappropriate words. ` +
                  `If you believe this is a mistake or have any queries, please email ${appealEmail}.`;
                await dmChannel.send(kickMessage);
              }
            } catch (error) {
              console.error('Error sending DM to kicked user:', error);
            }

            // Kick the user from the server
            try {
              await member.kick('Inappropriate nickname detected');
              console.log(`Kicked ${member.user.tag} from ${guild.name} for having an inappropriate nickname.`);
            } catch (error) {
              if (error.code === 50013) {
                console.error(`Missing permissions to kick ${member.user.tag} from ${guild.name}`);
              } else {
                console.error(`Error kicking ${member.user.tag} from ${guild.name}:`, error);
              }
            }
          }
        }
      } catch (error) {
        console.error(`Error fetching members from ${guild.name}:`, error);
      }
    }
  } catch (error) {
    console.error('Error fetching guilds:', error);
  }
}

// Kick users with inappropriate names on a timer (3 minutes)
setInterval(kickUsersWithInappropriateNames, 3 * 60 * 1000);

client.on('messageCreate', async (message) => {
  if (!message.content.startsWith(commandPrefix) || message.author.bot) {
    // Check for blocked words
    const lowerCaseContent = message.content.toLowerCase();
    const blockedWordFound = blockedWords.some((word) => lowerCaseContent.includes(word));

    if (blockedWordFound) {
      // Create an embed to log the offensive message
const embed = new MessageEmbed()
  .setColor('#FF0000')
  .setTitle('Offensive Message Logged')
  .addField('User ID', message.author.id)
  .addField('Username', message.author.username)
  .addField('Server Name', message.guild?.name || 'Direct Message')
  .addField('Message Content', message.content)
  .setFooter({ text: `Message ID: ${message.id}` }) // Update this line
  .setTimestamp();

      // Send the embed to the log channel
      const logChannel = client.channels.cache.get(logChannelId);
      if (logChannel && logChannel.isText()) {
        try {
          await logChannel.send({ embeds: [embed] });
        } catch (error) {
          console.error('Error sending the log message:', error);
        }
      }
    }
    return;
    }

  const args = message.content.slice(commandPrefix.length).trim().split(/ +/);
  const command = args.shift().toLowerCase();

  if (command === 'ban') {
    if (message.author.id !== allowedUserId) {
      return message.reply('You don\'t have permission to use this command.');
    }

    const userId = getUserIdFromMessage(args[0]);
    const banReason = args.slice(1).join(' ');

    if (!userId) {
      return message.reply('Please provide a valid user ID or mention to ban.');
    }

    try {
      const bannedUser = await client.users.fetch(userId);

      try {
        const dmChannel = await bannedUser.createDM();
        if (dmChannel) {
          const banMessage = `You have been banned from all servers which this bot is currently in for ${banReason || 'No reason provided'}. You may appeal to ${appealEmail}.`;
          await dmChannel.send(banMessage);
        } else {
          console.log(`Failed to send DM to banned user: ${bannedUser.tag}`);
        }
      } catch (error) {
        console.error('Error sending DM to banned user:', error);
      }

      const guilds = client.guilds.cache.values();
      for (const guild of guilds) {
        try {
          const member = await guild.members.fetch(userId);
          await member.ban({ reason: banReason });
          console.log(`Banned ${bannedUser.tag} from ${guild.name}`);
        } catch (error) {
          console.error(`Error banning ${bannedUser.tag} from ${guild.name}:`, error);
        }

        bannedUsers.set(userId, {
          username: bannedUser.username,
          userId: bannedUser.id,
          reason: banReason || 'No reason provided',
        });
      }

      saveBannedUsersToFile();
      message.reply('User has been banned from all servers.');
    } catch (error) {
      console.error('Error banning user:', error);
      message.reply('An error occurred while banning the user.');
    }
  } else if (command === 'unban') {
    if (message.author.id !== allowedUserId) {
      return message.reply('You don\'t have permission to use this command.');
    }

    const userId = getUserIdFromMessage(args[0]);

    if (!userId) {
      return message.reply('Please provide a valid user ID to unban.');
    }

    try {
      await unbanUserFromAllServers(userId);

      bannedUsers.delete(userId);

      saveBannedUsersToFile();
      refreshBannedUsersCache();

      message.reply('User has been unbanned from all servers.');
    } catch (error) {
      console.error('Error unbanning user:', error);
      message.reply('An error occurred while unbanning the user.');
    }
  } else if (command === 'bans') {
    if (message.author.id !== allowedUserId) {
      return message.reply('You don\'t have permission to use this command.');
    }

    if (bannedUsers.size === 0) {
      return message.reply('No users are currently banned.');
    }

    const page = parseInt(args[0], 10) || 1;

    const startIndex = (page - 1) * bansPerPage;
    const endIndex = page * bansPerPage;

    const bansArray = Array.from(bannedUsers.values());
    const totalPages = Math.ceil(bansArray.length / bansPerPage);

    if (page > totalPages) {
      return message.reply(`Invalid page number. Please enter a number between 1 and ${totalPages}.`);
    }

    const bansToDisplay = bansArray.slice(startIndex, endIndex);

    const embed = new MessageEmbed()
      .setColor('#FF0000')
      .setTitle('Banned Users')
      .setDescription(`Page ${page}`);

    bansToDisplay.forEach((banInfo) => {
      embed.addField('Username', banInfo.username, true);
      embed.addField('User ID', banInfo.userId, true);
      embed.addField('Reason', banInfo.reason, false);
      embed.addFields({ name: '\u200B', value: '\u200B' });
    });

    message.channel.send({ embeds: [embed] });
  } else if (command === 'help') {
    const botUser = client.user;
    const helpEmbed = new MessageEmbed()
      .setColor('#00FF00')
      .setTitle('Bot Help')
      .setDescription('This bot can perform the following commands:')
      .addField('Command: ban', 'Usage: `.ban <user> <reason>`\nDescription: Bans a user from all servers the bot is in.')
      .addField('Command: unban', 'Usage: `.unban <user>`\nDescription: Unbans a user.')
      .addField('Command: bans', 'Usage: `.bans [page]`\nDescription: Lists banned users.')
      .addField('Command: help', 'Usage: `.help`\nDescription: Displays this help message.');

    message.channel.send({ embeds: [helpEmbed] });
    
  } else if (command === 'servers') {
    if (message.author.id !== allowedUserId) {
      return message.reply('You don\'t have permission to use this command.');
    }

    const guilds = Array.from(client.guilds.cache.values());
    const page = parseInt(args[0], 10) || 1;
    const serversPerPage = 5;

    const totalPages = Math.ceil(guilds.length / serversPerPage);
    if (page < 1 || page > totalPages) {
      return message.reply(`Invalid page number. Please enter a number between 1 and ${totalPages}.`);
    }

    const startIndex = (page - 1) * serversPerPage;
    const endIndex = page * serversPerPage;
    const pagedGuilds = guilds.slice(startIndex, endIndex);

    const embed = new MessageEmbed()
      .setColor('#00FF00')
      .setTitle('List of Servers')
      .setDescription(`Page ${page}/${totalPages}\nHere are the servers:`)
      .setTimestamp();

    for (const guild of pagedGuilds) {
      embed.addFields(
        { name: 'Server Name', value: guild.name, inline: true },
        { name: 'Server ID', value: guild.id, inline: true },
        { name: 'Member Count', value: guild.memberCount.toString(), inline: true },
      );
    }

    message.channel.send({ embeds: [embed] });
  } else if (command === 'invite') {
    if (message.author.id !== allowedUserId) {
      return message.reply('You don\'t have permission to use this command.');
    }

    const userId = getUserIdFromMessage(args[0]);
    if (!userId) {
      return message.reply('Please provide a valid user ID to generate an invite.');
    }

    try {
      const invite = await message.channel.createInvite({ targetUser: userId });
      message.reply(`Here's the invite for the user with ID ${userId}: ${invite.url}`);
    } catch (error) {
      console.error('Error creating invite:', error);
      message.reply('An error occurred while generating the invite.');
    }
  }
});

client.on('guildMemberAdd', (member) => {
  if (bannedUsers.has(member.user.id)) {
    member.ban({ reason: 'Banned user rejoined the server.' })
      .then(() => {
        console.log(`Banned ${member.user.tag} who rejoined the server.`);
      })
      .catch((error) => {
        console.error('Error banning rejoined user:', error);
      });
  }
});

client.on('guildMemberAdd', async (member) => {
  const accountAgeInDays = (Date.now() - member.user.createdTimestamp) / (1000 * 60 * 60 * 24);

  if (accountAgeInDays < 7) {
    try {
      // Send a direct message to the user
      const dmChannel = await member.user.createDM();
      if (dmChannel) {
        const kickMessage =
          `Hello ${member.user.username},\n` +
          `You have been kicked from this server because your Discord account is less than 7 days old. ` +
          `If you believe this is a mistake, please email your appeal to ${appealEmail}.`;
        await dmChannel.send(kickMessage);
      }
    } catch (error) {
      console.error('Error sending DM to kicked user:', error);
    }

    // Kick the user from the server
    try {
      await member.kick('Account age is less than 7 days');
      console.log(`Kicked ${member.user.tag} from ${member.guild.name}`);
    } catch (error) {
      console.error(`Error kicking ${member.user.tag} from ${member.guild.name}:`, error);
    }
  }
});


// Register slash commands in every guild the bot is a member of
client.on('ready', async () => {
  try {
    const guilds = client.guilds.cache.values();
    for (const guild of guilds) {
      await guild.commands.set([
        {
          name: 'report',
          description: 'Report a user',
          options: [
            {
              name: 'user_id',
              description: 'User ID',
              type: 'STRING',
              required: true,
            },
            {
              name: 'username',
              description: 'Username',
              type: 'STRING',
              required: true,
            },
            {
              name: 'reporting_message',
              description: 'Reporting message',
              type: 'STRING',
              required: true,
            },
            {
              name: 'evidence',
              description: 'Screenshot or evidence',
              type: 'STRING',
              required: true,
            },
            {
              name: 'contact',
              description: 'Contact information (optional)',
              type: 'STRING',
              required: false,
            },
          ],
        },
        {
          name: 'help',
          description: 'Get bot help',
        },
      ]);
      console.log(`Slash commands registered in ${guild.name}`);
    }
    console.log('Slash commands registered successfully in all guilds.');
  } catch (error) {
    console.error('Error registering slash commands:', error);
  }
});

client.on('guildCreate', async (guild) => {
  try {
    await guild.commands.set([
      {
        name: 'report',
        description: 'Report a user',
        options: [
          {
            name: 'user_id',
            description: 'User ID',
            type: 'STRING',
            required: true,
          },
          {
            name: 'username',
            description: 'Username',
            type: 'STRING',
            required: true,
          },
          {
            name: 'reporting_message',
            description: 'Reporting message',
            type: 'STRING',
            required: true,
          },
          {
            name: 'evidence',
            description: 'Screenshot or evidence',
            type: 'STRING',
            required: true,
          },
          {
            name: 'contact',
            description: 'Contact information (optional)',
            type: 'STRING',
            required: false,
          },
        ],
      },
      {
        name: 'help',
        description: 'Get bot help',
      },
    ]);
    console.log(`Slash commands registered in ${guild.name}`);

    // Check for banned users in the newly joined server
    const bans = await guild.bans.fetch();
    for (const [userId, banInfo] of bannedUsers) {
      if (bans.has(userId)) {
        const bannedUser = bans.get(userId).user;
        try {
          await guild.bans.remove(userId, 'Banned user detected in the server.');
          console.log(`Banned user with ID ${userId} (${bannedUser.tag}) from ${guild.name}`);
        } catch (error) {
          console.error(`Error banning user with ID ${userId} (${bannedUser.tag}) from ${guild.name}:`, error);
        }
      }
    }
  } catch (error) {
    console.error(`Error registering slash commands in ${guild.name}:`, error);
  }
});

// Handle slash command interactions
client.on('interactionCreate', async (interaction) => {
  if (!interaction.isCommand()) return;

  const { commandName, options } = interaction;

  if (commandName === 'report') {
    const userId = options.getString('user_id');
    const username = options.getString('username');
    const reportingMessage = options.getString('reporting_message');
    const evidence = options.getString('evidence');
    const contact = options.getString('contact');

    if (!userId || !username || !reportingMessage || !evidence) {
      return interaction.reply('Please provide all required fields.');
    }

    const reportChannelId = 'report channel ID'; // Replace with the ID of the report channel

    const reportChannel = client.channels.cache.get(reportChannelId);
    if (!reportChannel || reportChannel.type !== 'GUILD_TEXT') {
      return interaction.reply('Failed to find the report channel. Please contact the bot administrator.');
    }

    let reportedUser;
    try {
      reportedUser = await client.users.fetch(userId);
    } catch (error) {
      console.error('Error fetching user:', error);
      return interaction.reply('Failed to fetch user information.');
    }

    const embed = new MessageEmbed()
      .setColor('#FF0000')
      .setTitle('New Report')
      .addField('User ID', userId)
      .addField('Username', username)
      .addField('Reporting Message', reportingMessage)
      .addField('Evidence', evidence) // Added evidence field
      .addField('Contact', contact || 'Not provided')
      .addField('Reported by', `${interaction.user.tag} (${interaction.user.id})`)
      .setFooter(`Reported by ${interaction.user.tag}`)
      .setTimestamp();

    try {
      await reportChannel.send({ embeds: [embed] });
      await interaction.reply('Report submitted successfully.');
    } catch (error) {
      console.error('Error sending report:', error);
      await interaction.reply('Failed to submit the report. Please try again later.');
    }
  } else if (commandName === 'help') {
    const helpEmbed = new MessageEmbed()
      .setColor('#00FF00')
      .setTitle('ModGuardian Help')
      .setDescription('ModGuardian is a powerful moderation bot designed to keep your server safe. It will automatically ban any users which has been reported to us via the report command.')
      .addFields(
        { name: 'Command: report', value: 'Usage: `/report <user_id> <username> <reporting_message> <evidence> [contact]`', inline: false },
        { name: 'Description:', value: 'Report a user to the moderation team.', inline: true },
        { name: 'Command: help', value: 'Usage: `/help`', inline: false },
        { name: 'Description:', value: 'Displays this help message.', inline: true }
      )
      .setFooter('For further assistance, contact ModGuardian Support');

    await interaction.reply({ embeds: [helpEmbed] });
  }
});

client.login('BOT TOKEN'); //Replace with your bot token

File: bot.py
import discord
from discord import Embed  
import openai
from discord.ext import commands, tasks
from datetime import datetime
import re

# Discord Bot Token
DISCORD_TOKEN = 'BOT_TOKEN'

# OpenAI API Key (text moderation costs $0 to run)
OPENAI_API_KEY = 'OPENAI_API_KEY'


LOG_CHANNEL_ID = 1150900821693833366  # Replace with your Discord channel ID of logging channel
AUTHORIZED_USER_IDS = ['974368593615659098', '']  # Replace with actual user IDs of mods
BAN_APPEAL_EMAIL = 'appeals@bytelabs.site'  # Replace with your actual email for appeals
BANNED_USERS_FILE = 'banned_users.txt'  # Path to the file where banned user info will be stored
WARNINGS_FILE = 'warnings.txt' # Path to the file where warnings are stored for record keeping 
EXCLUDED_CATEGORIES = ['self-harm', 'self-harm/intent', 'self-harm/instructions'] # What categories won't be moderated

# Define the intents.
intents = discord.Intents.default()
intents.messages = True
intents.guilds = True
intents.message_content = True
intents.dm_messages = True
intents.members = True

# Initialize Discord Bot with intents
bot = commands.Bot(command_prefix='...', intents=intents, help_command=None)

# OpenAI client setup
openai.api_key = OPENAI_API_KEY

# Interval for nickname checks in minutes
NICKNAME_CHECK_INTERVAL = 10  # e.g., check every 10 minutes

def get_user_id(mention):
    return int(re.sub(r'[<@!>]', '', mention)) if mention.startswith('<@') and mention.endswith('>') else int(mention)
    
    
@bot.command()
async def search(ctx, user_mention):
    if str(ctx.author.id) not in AUTHORIZED_USER_IDS:
        await ctx.send('You do not have permission to use this command.')
        return

    try:
        # Try converting mention or ID to an integer
        user_id = get_user_id(user_mention)

        # Check if the user ID corresponds to a real user in Discord
        user = bot.get_user(user_id)
        if not user:
            await ctx.send("Invalid user or user ID.")
            return

        # Search in banned users file
        banned_info = "Not banned."
        with open(BANNED_USERS_FILE, 'r') as f:
            for line in f:
                if str(user_id) in line:
                    banned_info = line.strip()
                    break

        # Search in warnings file
        warnings_info = []
        with open(WARNINGS_FILE, 'r') as f:
            for line in f:
                if str(user_id) in line:
                    warnings_info.append(line.strip())

        warning_count = len(warnings_info)
        warnings_text = "\n".join(warnings_info) if warnings_info else "No warnings."

        
        embed = discord.Embed(title=f"Search Results for {user}", color=0x00ff00)
        embed.add_field(name="Banned Info", value=banned_info, inline=False)
        embed.add_field(name=f"Warnings ({warning_count})", value=warnings_text, inline=False)
        await ctx.send(embed=embed)

    except ValueError:
        await ctx.send("Invalid user or user ID format. Please use a valid mention or numeric ID.")


@bot.command()
async def warn(ctx, member, *, reason=None):
    # Authorization check
    if str(ctx.author.id) not in AUTHORIZED_USER_IDS:
        await ctx.send('You do not have permission to use this command.')
        return

    # Convert member mention to ID or use the provided ID
    member_id = ''.join(filter(str.isdigit, member))

    # Attempt to convert member_id to int
    try:
        member_id = int(member_id)
    except ValueError:
        await ctx.send("Invalid user ID.")
        return

    # Attempt to fetch the user
    try:
        user = await bot.fetch_user(member_id)
        dm_message = f"You have been warned for the following reason: {reason}"
        await user.send(dm_message)
    except discord.NotFound:
        await ctx.send(f"User with ID {member_id} not found.")
        return
    except discord.HTTPException:
        await ctx.send(f"Failed to send DM to user with ID {member_id}.")

    # Log the warning to a file
    with open(WARNINGS_FILE, 'a') as file:
        file.write(f"{member_id}, {reason}\n")

    # Notify the command issuer
    await ctx.send(f"User with ID {member_id} has been warned.")

@bot.command()
async def ban(ctx, member_id: str, *, reason=None):
    # Check if the author is allowed to use the command
    if str(ctx.author.id) not in AUTHORIZED_USER_IDS:
        await ctx.send("You are not authorized to use this command.")
        return

    # Convert member_id to an integer if it's a mention or direct ID
    member_id = int(''.join(filter(str.isdigit, member_id)))

    # Add the user to the banned users list
    with open(BANNED_USERS_FILE, 'a') as file:
        file.write(f"{member_id}, {reason}\n")
    
    # Attempt to fetch the user
    try:
        user = await bot.fetch_user(member_id)
        dm_message = f"You have been banned for the following reason: {reason}\nYou may appeal this ban at {BAN_APPEAL_EMAIL}."
        await user.send(dm_message)
    except discord.NotFound:
        pass  # User not found, possibly due to an invalid ID
    except discord.HTTPException:
        pass  # Failed to send DM, possibly due to blocked DMs

    # Ban the user from all servers where the bot is present
    for guild in bot.guilds:
        member = guild.get_member(member_id)
        if member:
            try:
                await guild.ban(member, reason=reason)
            except discord.Forbidden:
                pass  # Bot doesn't have ban permissions in this guild

    await ctx.send(f"User with ID {member_id} has been banned.")

# Function to check content with OpenAI's Moderation API
async def check_content_for_inappropriateness(content):
    try:
        response = openai.Moderation.create(
            input=content,
            model="text-moderation-latest"
        )
        flagged = response['results'][0]['flagged']
        categories = response['results'][0]['categories']
        category_scores = response['results'][0]['category_scores']
        return flagged, categories
    except Exception as e:
        print(f"Error checking content: {e}")
        return False, {}

# Function to create an embed for logging
def create_log_embed(message, categories):
    embed = discord.Embed(title="Inappropriate message detected", color=0xFF5733)  # Red color for violations
    embed.add_field(name="User ID", value=message.author.id, inline=False)
    embed.add_field(name="Username", value=str(message.author), inline=False)  # Username with discriminator
    embed.add_field(name="Message ID", value=message.id, inline=False)
    embed.add_field(name="Server Name", value=message.guild.name, inline=False)
    embed.add_field(name="Server ID", value=message.guild.id, inline=False)
    embed.add_field(name="Message Content", value=message.content, inline=False)
    
    # Format the moderation categories for neatness
    categories_formatted = '\n'.join(f"{category}: {score:.2f}" for category, score in categories.items())
    embed.add_field(name="Moderation Categories", value=categories_formatted, inline=False)
    
    embed.set_footer(text=message.created_at.strftime("%Y-%m-%d %H:%M:%S"))
    return embed

# Check and take action on inappropriate username or nickname
async def handle_inappropriate_member(member):
    if member == bot.user or member.top_role >= member.guild.me.top_role:
        return

    name_to_check = member.display_name
    flagged, scores = await check_content_for_inappropriateness(name_to_check)

    excluded_categories = ['self-harm', 'self-harm/intent', 'self-harm/instructions']
    action_needed = any(category for category, score in scores.items() if score > 0.8 and category not in excluded_categories)

    if flagged and action_needed:
        try:
            dm_message = "Your username or nickname has been flagged as inappropriate. Please change it to comply with our server's guidelines."
            await member.send(dm_message)
            await member.kick(reason="Inappropriate username or nickname detected by moderation model.")
            
            log_channel = bot.get_channel(LOG_CHANNEL_ID)
            if log_channel:
                embed = discord.Embed(title="Content Violation", color=0xFF0000)
                embed.add_field(name="User ID", value=member.id, inline=True)
                embed.add_field(name="Username", value=str(member), inline=True)
                embed.add_field(name="Server ID", value=member.guild.id, inline=True)
                embed.add_field(name="Server Name", value=member.guild.name, inline=True)
                embed.add_field(name="Violating Name", value=name_to_check, inline=True)
                embed.add_field(name="OpenAI Moderation Values", value=str(scores), inline=False)
                embed.set_footer(text=datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
                await log_channel.send(embed=embed)
        except discord.Forbidden:
            # Optionally, you can log these messages if needed for debugging
            # print(f"Bot does not have permission to kick or send DM to {member.display_name} (ID: {member.id}).")
            pass
        except discord.HTTPException as e:
            # Optionally, you can log these messages if needed for debugging
            # print(f"Error handling member {member.display_name} (ID: {member.id}): {e}")
            pass
        
# Function to check all member nicknames in a server
async def check_all_members(guild):
    for member in guild.members:
        if member.nick or member.name:
            await handle_inappropriate_member(member)

# Task loop to periodically check nicknames
@tasks.loop(minutes=NICKNAME_CHECK_INTERVAL)
async def check_nicknames_periodically():
    for guild in bot.guilds:
        await check_all_members(guild)
        
@bot.command()
async def unban(ctx, member_id: str):
    # Check if the author is allowed to use the command
    if str(ctx.author.id) not in AUTHORIZED_USER_IDS:
        await ctx.send("You are not authorized to use this command.")
        return

    # Convert member_id to an integer if it's a mention or direct ID
    member_id = int(''.join(filter(str.isdigit, member_id)))

    # Read the current banned users list
    with open(BANNED_USERS_FILE, 'r') as file:
        banned_users = file.readlines()

    # Check if the user is in the banned users list
    if not any(str(member_id) in line for line in banned_users):
        await ctx.send(f"User with ID {member_id} is not in the banned users list.")
        return

    # Remove the user from the banned users list
    with open(BANNED_USERS_FILE, 'w') as file:
        for line in banned_users:
            if str(member_id) not in line:
                file.write(line)

    # Unban the user from all servers where the bot is present
    for guild in bot.guilds:
        try:
            user = await bot.fetch_user(member_id)
            await guild.unban(user)
        except discord.NotFound:
            pass  # User not banned in this server
        except discord.Forbidden:
            pass  # Bot doesn't have unban permissions in this server

    # Notify the command issuer
    await ctx.send(f"User with ID {member_id} has been unbanned from all servers.")
        
@bot.event
async def on_member_join(member):
    with open(BANNED_USERS_FILE, 'r') as file:
        banned_users = file.readlines()

    # Check if the joining member is in the banned users list
    if any(str(member.id) in line for line in banned_users):
        try:
            await member.guild.ban(member, reason="Banned user rejoined.")
            dm_message = f"You are banned. You may appeal this ban at {BAN_APPEAL_EMAIL}."
            await member.send(dm_message)
        except discord.Forbidden:
            pass  # Bot doesn't have ban permissions
        except discord.HTTPException:
            pass  # Failed to send DM or ban user
    
@bot.event
async def on_guild_join(guild):
    await check_all_members(guild)

@tasks.loop(minutes=NICKNAME_CHECK_INTERVAL)
async def check_nicknames_periodically():
    for guild in bot.guilds:
        await check_all_members(guild)
        
@check_nicknames_periodically.before_loop
async def before_check_nicknames_periodically():
    await bot.wait_until_ready()
    
# Start the nickname check loop when bot is ready
@bot.event
async def on_ready():
    print(f'Logged in as {bot.user}')
    if not check_nicknames_periodically.is_running():
        check_nicknames_periodically.start()
        
@bot.event
async def on_disconnect():
    if check_nicknames_periodically.is_running():
        check_nicknames_periodically.cancel()

@bot.event
async def on_error(event_method, *args, **kwargs):
    with open('err.log', 'a') as f:
        if event_method == 'on_message':
            f.write(f'Unhandled message: {args[0]}\n')
        else:
            raise
# Event listener for on_message
@bot.event
async def on_message(message):
    # Ignore bot's own messages and other bots
    if message.author.bot:
        return

    try:
        flagged, categories = await check_content_for_inappropriateness(message.content)
        if flagged:
            log_channel = bot.get_channel(LOG_CHANNEL_ID)
            embed = create_log_embed(message, categories)
            await log_channel.send(embed=embed)

            # Delete message and DM user if the confidence is 100% for any category except self-harm & self-harm/intent
            should_delete = any(score == 1.0 for cat, score in categories.items() if cat not in ['self-harm', 'self-harm/intent', 'self-harm/instructions'])
            if should_delete:
                # Delete the message
                await message.delete()
                # Send a DM to the user
                dm_message = "Your message was deleted because it violated discord's community guidelines."
                await message.author.send(dm_message)
            else:
                # Handle self-harm messages by sending a supportive DM
                if categories.get('self-harm/intent', 0) >= 0.9 or categories.get('self-harm/instructions', 0) >= 0.9:
                    dm_message = "Hey there! We recently flagged a message concerning self harm about yourself or someone else. If you are struggling with mental health please feel free to reach out to someone trusted and a professionally licensed mental health professional and/or a hotline or charity. We are here for you and as a result we forward all concerning messages about harm against themselves and someone else to discord due to our policies. Please contact @Kurope on discord if you have any further questions or concerns or feel like you want to talk to someone!"
                    await message.author.send(dm_message)

    except Exception as e:
        print(f"Error processing message: {e}")

    
    await bot.process_commands(message)


@bot.group(invoke_without_command=True)
async def help(ctx):
    bot_description = (
        "Modguardian is a moderation assistant designed to maintain community guidelines "
        "and ensure a safe environment for all members. It uses advanced AI to monitor "
        "messages, detect inappropriate content, and take necessary actions such as warnings "
        "or bans. Below is a list of commands that authorized users (bot owner) can use to manage "
        "community moderation. If banned with ModGuardian the advanced AI system will keep you banned from every server you share with the bot."
    )
    embed = discord.Embed(title="Modguardian Help", description=bot_description, color=discord.Color.green())
    embed.add_field(name="!search <userID or @user>", value="**Admin use only** - Searches a user's punishments and if they are in our system", inline=False)
    embed.add_field(name="!ban <userID or @user> <reason>", value="**Admin use only** - Bans a user across all servers where the bot is present", inline=False)
    embed.add_field(name="!warn <userID or @user> <reason>", value="**Admin use only** - Warns a user and records the warning in the system", inline=False)
    await ctx.send(embed=embed)
    
# Start the bot
bot.run(DISCORD_TOKEN)
